{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import re\n",
    "##10/02/23: this was my first iteration but as you can see its very complicated and quite a rabbit hole. removing all null genotyped variants helps to sift all of these possible scenarios out. IGNORE THIS CELL\n",
    "\n",
    "\n",
    "## we are trying to keep all variants that can be inherited in autosomal recessive fashion(parents heterozygous or homozygous ref) as well as including the possibility of de novo variants\n",
    "with open('/Users/johnlee/jupyter/new_case/high_mod.qual100.C4RCD_0637_1_PB_Whole_F0020G2S2M0635P0636_TSE61-C4RCD_0635_1_PB_Whole_F0020G2S1MP_TSE61-C4RCD_0636_1_PB_Whole_F0020G1S1MP_TSE61.phaseBT.vcf', 'rt') as vcf, open(\"inherited_possible_denovo.vcf\", \"w\") as output:\n",
    "    \n",
    "    for line in vcf:\n",
    "        if line.startswith(\"#\"):\n",
    "            output.write(line)\n",
    "        else:\n",
    "            regex = \"GT:\\S*?\\s+(.+?)[\\/|](.+?)\\S*?\\s+(.+?)[\\/|](.+?)\\S*?\\s+(.+?)[\\/|](.+?)\"\n",
    "            genotype = re.search(regex, line)\n",
    "            ## remove all variants with no information on child\n",
    "            if \".\" in genotype[5] or \".\" in genotype[6]:\n",
    "                print('null genotype for child')\n",
    "\n",
    "            elif \".\" in genotype[1] or \".\" in genotype[2]:\n",
    "                ## if both parents are null, it means kid either has reference or alt allele; keep it. it has to be a  DE NOVO VARIANT. filter for alternative alleles only, as reference alleles arent(???) considered pathogenic.\n",
    "                if \".\" in genotype[3] or \".\" in genotype[4]:\n",
    "                    if float(genotype[5]) > 0 or float(genotype[6]) > 0:\n",
    "                        output.write(line)\n",
    "                else: ## IF PARENT 1 IS NULL, check inheritance of PARENT 2;if one of the parents is null but the other one isnt,homozygous parents cannot be giving their alleles to the kid(parents cant have the disease). keep lines of which the alleles the kid has, are not of the parent\n",
    "                    if genotype[3] != genotype[5] or genotype[3] != genotype[6] or genotype[4] != genotype[5] or genotype[4] != genotype[5]:\n",
    "                        ## same thing but parent 2\n",
    "                        output.write(line)\n",
    "                    ## scenario where parent contributes one alternate allele, and kid receives the same alt allele de novo    \n",
    "                    elif float(genotype[3]) > 0 and genotype[3] == genotype[5] and genotype[3] == genotype[6]:\n",
    "                        output.write(line)\n",
    "                    elif float(genotype[4]) > 0 and genotype[4] == genotype[5] and genotype[4] == genotype[6]:\n",
    "                        output.write(line)\n",
    "\n",
    "            ## if parent 2 is null but parent 1 isn't, keep the line. \n",
    "            elif \".\" in genotype[3] or \".\" in genotype[4]:\n",
    "                if genotype[1] != genotype[5] or genotype[1] != genotype[6] or genotype[2] != genotype[5] or genotype[2] != genotype[5]:\n",
    "                        ## same thing but parent 2\n",
    "                    output.write(line)\n",
    "                    ## scenario where parent contributes one alternate allele, and kid receives the same alt allele de novo    \n",
    "                elif float(genotype[1]) > 0 and genotype[1] == genotype[5] and genotype[1] == genotype[6]:\n",
    "                    output.write(line)\n",
    "                elif float(genotype[2]) > 0 and genotype[2] == genotype[5] and genotype[2] == genotype[6]:\n",
    "                    output.write(line)\n",
    "\n",
    "\n",
    "            ## checking homozygosity/alternate status of parent 1. if they are homozygous alternate, then dont keep, UNLESS the next if statement occurs(ive written it backwards)\n",
    "            ##THIS IS FOR IF BOTH PARENTS ARENT NULL\n",
    "\n",
    "            elif genotype[1] == genotype[2] and float(genotype[1]) > 0:\n",
    "                ## this step is for the scenario of kid having different alternate allele than parent1. if either allele is different than parent1's allele, then keep it. parent1's alternate may not be pathogenic, the other one may be\n",
    "                if genotype[1] != genotype[5] or genotype[1] != genotype[6]:\n",
    "                    output.write(line)\n",
    "            ## checking homozygosity/alternate status of parent 2\n",
    "            elif genotype[3] == genotype[4] and float(genotype[3]) > 0:\n",
    "                ## same thing but parent 2\n",
    "                if genotype[3] != genotype[5] or genotype[3] != genotype[6]:\n",
    "                    output.write(line)\n",
    "            else:\n",
    "                output.write(line)\n",
    "\n",
    "        \n",
    "        \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import re\n",
    "\n",
    "## we are trying to keep all variants that can be inherited in autosomal recessive fashion(parents heterozygous or homozygous ref) as well as including the possibility of de novo variants. remove all null genotyped variants\n",
    "with open('/Users/johnlee/jupyter/new_case/high_mod.qual100.C4RCD_0637_1_PB_Whole_F0020G2S2M0635P0636_TSE61-C4RCD_0635_1_PB_Whole_F0020G2S1MP_TSE61-C4RCD_0636_1_PB_Whole_F0020G1S1MP_TSE61.phaseBT.vcf', 'rt') as vcf, open(\"no_null.high_mod.qual100.C4RCD_0637_1_PB_Whole_F0020G2S2M0635P0636_TSE61-C4RCD_0635_1_PB_Whole_F0020G2S1MP_TSE61-C4RCD_0636_1_PB_Whole_F0020G1S1MP_TSE61.phaseBT.vcf\", \"w\") as output:\n",
    "    \n",
    "    for line in vcf:\n",
    "        if line.startswith(\"#\"):\n",
    "            output.write(line)\n",
    "        else:\n",
    "            regex = \"GT:\\S*?\\s+(.+?)[\\/|](.+?)\\S*?\\s+(.+?)[\\/|](.+?)\\S*?\\s+(.+?)[\\/|](.+?)\"\n",
    "            genotype = re.search(regex, line)\n",
    "            ## remove all variants with no information on child\n",
    "            if \".\" in genotype[1] or \".\" in genotype[2] or \".\" in genotype[3] or \".\" in genotype[4] or \".\" in genotype[5] or \".\" in genotype[6]:\n",
    "                print('null genotype for child')\n",
    "            else:\n",
    "                output.write(line)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "## MENDELIAN INHERITANCE VARIANTS\n",
    "## WE NEED TO CHECK FOR JUST INHERITED VARIANTS, MEANING PARENTS MUST BE HETEROZYGOUS AND THE ALLELES THE KID HAS MUST BE FROM PARENTS\n",
    "import re\n",
    "\n",
    "## we are trying to keep all variants that can be inherited in autosomal recessive fashion(parents heterozygous or homozygous ref) as well as including the possibility of de novo variants\n",
    "with open('/Users/johnlee/jupyter/new_case/high_mod.qual100.C4RCD_0637_1_PB_Whole_F0020G2S2M0635P0636_TSE61-C4RCD_0635_1_PB_Whole_F0020G2S1MP_TSE61-C4RCD_0636_1_PB_Whole_F0020G1S1MP_TSE61.phaseBT.vcf', 'rt') as vcf, open(\"inherited_possible_denovo.vcf\", \"w\") as output:\n",
    "    \n",
    "    for line in vcf:\n",
    "        if line.startswith(\"#\"):\n",
    "            output.write(line)\n",
    "        else:\n",
    "            regex = \"GT:\\S*?\\s+(.+?)[\\/|](.+?)\\S*?\\s+(.+?)[\\/|](.+?)\\S*?\\s+(.+?)[\\/|](.+?)\"\n",
    "            genotype = re.search(regex, line)\n",
    "            ## remove all variants with no information on child\n",
    "            if \".\" in genotype[5] or \".\" in genotype[6]:\n",
    "                print('null genotype for child')\n",
    "            elif \".\" in genotype[1] or \".\" in genotype[2]:\n",
    "                if \".\" in genotype[3] or \".\" in genotype[4]:\n",
    "                    ##neither parents have reported variant here. since we are only looking for inherited variants here, discard them\n",
    "                    print('no inheritance')\n",
    "                \n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "## DENOVO VARIANTS HERE \n",
    "\n",
    "with open('qual100.C4RCD_0637_1_PB_Whole_F0020G2S2M0635P0636_TSE61-C4RCD_0635_1_PB_Whole_F0020G2S1MP_TSE61-C4RCD_0636_1_PB_Whole_F0020G1S1MP_TSE61.phaseBT.vcf', 'rt') as vcf, open('denovo_only.vcf', 'w') as output:\n",
    "    regex = \"GT:\\S*?\\s+(.+?)[\\/|](.+?)\\S*?\\s+(.+?)[\\/|](.+?)\\S*?\\s+(.+?)[\\/|](.+?)\"\n",
    "    for line in vcf:\n",
    "        if line.startswith('#'):\n",
    "            output.write(line)\n",
    "        else:\n",
    "            genotype = re.search(regex, line)\n",
    "            parent_alleles = [genotype[1], genotype[2], genotype[3], genotype[4]]\n",
    "            ## if both alleles of kid are in \n",
    "            if \".\" in genotype[5] or \".\" in genotype[6]:\n",
    "                print('null genotype for child')\n",
    "            elif genotype[5] in parent_alleles and genotype[6] in parent_alleles:\n",
    "                print('not denovo')\n",
    "            else:\n",
    "                ## assumptoion here: that the de novo variant is going to be a single hit. it seems like the odds are astronomically low, but considering this kid's disease is super rare, we may not be exempt from rare scenarios\n",
    "                if genotype[5] != genotype[6] and float(genotype[5]) > 0 and float(genotype[6]) > 0:\n",
    "                    output.write(line)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "## 230929. use the no null vcf as david mentioned we should exclude any variatns without complete information\n",
    "\n",
    "## here, we will filter for the de novo variants, dominant negative\n",
    "import re\n",
    "with open('/Users/johnlee/jupyter/new_case/case_study_trio/no_null.high_mod.qual100.C4RCD_0637_1_PB_Whole_F0020G2S2M0635P0636_TSE61-C4RCD_0635_1_PB_Whole_F0020G2S1MP_TSE61-C4RCD_0636_1_PB_Whole_F0020G1S1MP_TSE61.phaseBT.vcf', 'rt') as vcf, open('/Users/johnlee/jupyter/new_case/case_study_trio/denovo.no_null.high_mod.qual100.C4RCD_0637_1_PB_Whole_F0020G2S2M0635P0636_TSE61-C4RCD_0635_1_PB_Whole_F0020G2S1MP_TSE61-C4RCD_0636_1_PB_Whole_F0020G1S1MP_TSE61.phaseBT.vcf', 'w') as output:\n",
    "    regex = \"GT:\\S*?\\s+(.+?)[\\/|](.+?)\\S*?\\s+(.+?)[\\/|](.+?)\\S*?\\s+(.+?)[\\/|](.+?)\"\n",
    "\n",
    "    for line in vcf:\n",
    "        if line.startswith('#'):\n",
    "            output.write(line)\n",
    "        else:\n",
    "            genotype = re.search(regex, line)\n",
    "            parent_alleles = [genotype[1], genotype[2], genotype[3], genotype[4]]\n",
    "            if genotype[5] not in parent_alleles or genotype[6] not in parent_alleles:\n",
    "                output.write(line)\n",
    "\n",
    "        \n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import re\n",
    "import pandas as pd\n",
    "with open('/Users/johnlee/jupyter/new_case/case_study_trio/no_null.high_mod.qual100.C4RCD_0637_1_PB_Whole_F0020G2S2M0635P0636_TSE61-C4RCD_0635_1_PB_Whole_F0020G2S1MP_TSE61-C4RCD_0636_1_PB_Whole_F0020G1S1MP_TSE61.phaseBT.vcf', 'rt') as vcf, open('/Users/johnlee/jupyter/new_case/case_study_trio/aut_rec.no_null.high_mod.qual100.C4RCD_0637_1_PB_Whole_F0020G2S2M0635P0636_TSE61-C4RCD_0635_1_PB_Whole_F0020G2S1MP_TSE61-C4RCD_0636_1_PB_Whole_F0020G1S1MP_TSE61.phaseBT.vcf', 'w') as output:\n",
    "    regex = \"GT:\\S*?\\s+(.+?)[\\/|](.+?)\\S*?\\s+(.+?)[\\/|](.+?)\\S*?\\s+(.+?)[\\/|](.+?)\"\n",
    "    gene_list_regex = \"EFF=\\S*?[|]\\S*?[|]\\S*?[|]\\S*?[|]\\S*?[|](.*?)[|]\"\n",
    "    gene_list = []\n",
    "## first make a gene list \n",
    "    for line in vcf:\n",
    "        if not line.startswith('#'):\n",
    "            gene_name = re.search(gene_list_regex, line)\n",
    "            gene_list.append(gene_name[1])\n",
    "    \n",
    "    genes_df = pd.DataFrame(gene_list)\n",
    "    gene_counts = genes_df[0].value_counts().reset_index()\n",
    "    gene_counts.columns = ['gene', 'count']\n",
    "    two_hits = gene_counts.loc[gene_counts['count'] >1]\n",
    "\n",
    "    for line in vcf:\n",
    "        if line.startswith('#'):\n",
    "            output.write(line)\n",
    "        else:\n",
    "            line_gene = re.search(gene_list_regex, line)\n",
    "            if line_gene[1] in two_hits['gene']:\n",
    "                output.write(line)\n",
    "            else:\n",
    "                genotype = re.search(regex, line)\n",
    "                if genotype[5] == genotype[6]:\n",
    "                    parent1 = [genotype[1], genotype[2]]\n",
    "                    parent2 = [genotype[3], genotype[4]]\n",
    "                    if genotype[5] in parent1 and genotype[5] in parent2:\n",
    "                        output.write(line)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import re \n",
    "import pandas as pd\n",
    "\n",
    "##creating both aut_rec and comp_het:\n",
    "\n",
    "with open('/Users/johnlee/jupyter/new_case/case_study_trio/no_null.high_mod.qual100.C4RCD_0637_1_PB_Whole_F0020G2S2M0635P0636_TSE61-C4RCD_0635_1_PB_Whole_F0020G2S1MP_TSE61-C4RCD_0636_1_PB_Whole_F0020G1S1MP_TSE61.phaseBT.vcf', 'rt') as vcf, open('/Users/johnlee/jupyter/new_case/case_study_trio/aut_rec.no_null.high_mod.qual100.C4RCD_0637_1_PB_Whole_F0020G2S2M0635P0636_TSE61-C4RCD_0635_1_PB_Whole_F0020G2S1MP_TSE61-C4RCD_0636_1_PB_Whole_F0020G1S1MP_TSE61.phaseBT.vcf', 'w') as aut_rec,  open('/Users/johnlee/jupyter/new_case/case_study_trio/comp_het.no_null.high_mod.qual100.C4RCD_0637_1_PB_Whole_F0020G2S2M0635P0636_TSE61-C4RCD_0635_1_PB_Whole_F0020G2S1MP_TSE61-C4RCD_0636_1_PB_Whole_F0020G1S1MP_TSE61.phaseBT.vcf', 'w') as comp_het:\n",
    "    regex = \"GT:\\S*?\\s+(.+?)[\\/|](.+?)\\S*?\\s+(.+?)[\\/|](.+?)\\S*?\\s+(.+?)[\\/|](.+?)\"\n",
    "    gene_list_regex = \"EFF=\\S*?[|]\\S*?[|]\\S*?[|]\\S*?[|]\\S*?[|](.*?)[|]\"\n",
    "    gene_list = []\n",
    "## first make a gene list \n",
    "    for line in vcf:\n",
    "        if not line.startswith('#'):\n",
    "            gene_name = re.search(gene_list_regex, line)\n",
    "            gene_list.append(gene_name[1])\n",
    "    \n",
    "    genes_df = pd.DataFrame(gene_list)\n",
    "    gene_counts = genes_df[0].value_counts().reset_index()\n",
    "    gene_counts.columns = ['gene', 'count']\n",
    "    two_hits = gene_counts.loc[gene_counts['count'] >1]\n",
    "\n",
    "    vcf.seek(0)\n",
    "\n",
    "    for line in vcf:\n",
    "        if line.startswith('#'):\n",
    "            aut_rec.write(line)\n",
    "            comp_het.write(line)\n",
    "        else:\n",
    "            line_gene = re.search(gene_list_regex, line)\n",
    "            ## two hits to \n",
    "            if line_gene[1] in two_hits['gene'].values:\n",
    "                comp_het_re = re.search(regex, line)\n",
    "                parent1 = [genotype[1], genotype[2]]\n",
    "                parent2 = [genotype[3], genotype[4]]\n",
    "                if genotype[5] in parent1 or genotype[6] in parent1 and genotype[1] != genotype[2]:\n",
    "                    comp_het.write(line)\n",
    "                elif genotype[5] in parent2 or genotype[6] in parent2 and genotype[3] != genotype[4]:\n",
    "                    comp_het.write(line)\n",
    "            else:\n",
    "                genotype = re.search(regex, line)\n",
    "                if genotype[5] == genotype[6] and float(genotype[5]) > 0:\n",
    "                    parent1 = [genotype[1], genotype[2]]\n",
    "                    parent2 = [genotype[3], genotype[4]]\n",
    "                    if genotype[5] in parent1 and genotype[5] in parent2 and genotype[1] != genotype[2] and genotype[3] != genotype[4]:\n",
    "                        aut_rec.write(line)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "## for some reason i need to separate the list creation so oi copy pastted the vcf writing to another cell here\n",
    "\n",
    "with open('/Users/johnlee/jupyter/new_case/case_study_trio/no_null.high_mod.qual100.C4RCD_0637_1_PB_Whole_F0020G2S2M0635P0636_TSE61-C4RCD_0635_1_PB_Whole_F0020G2S1MP_TSE61-C4RCD_0636_1_PB_Whole_F0020G1S1MP_TSE61.phaseBT.vcf', 'rt') as vcf, open('/Users/johnlee/jupyter/new_case/case_study_trio/aut_rec.no_null.high_mod.qual100.C4RCD_0637_1_PB_Whole_F0020G2S2M0635P0636_TSE61-C4RCD_0635_1_PB_Whole_F0020G2S1MP_TSE61-C4RCD_0636_1_PB_Whole_F0020G1S1MP_TSE61.phaseBT.vcf', 'w') as aut_rec,  open('/Users/johnlee/jupyter/new_case/case_study_trio/comp_het.no_null.high_mod.qual100.C4RCD_0637_1_PB_Whole_F0020G2S2M0635P0636_TSE61-C4RCD_0635_1_PB_Whole_F0020G2S1MP_TSE61-C4RCD_0636_1_PB_Whole_F0020G1S1MP_TSE61.phaseBT.vcf', 'w') as comp_het:\n",
    "    regex = \"GT:\\S*?\\s+(.+?)[\\/|](.+?)\\S*?\\s+(.+?)[\\/|](.+?)\\S*?\\s+(.+?)[\\/|](.+?)\"\n",
    "    gene_list_regex = \"EFF=\\S*?[|]\\S*?[|]\\S*?[|]\\S*?[|]\\S*?[|](.*?)[|]\"\n",
    "    for line in vcf:\n",
    "        if line.startswith('#'):\n",
    "            aut_rec.write(line)\n",
    "            comp_het.write(line)\n",
    "        else:\n",
    "            line_gene = re.search(gene_list_regex, line)\n",
    "            ## two hits to \n",
    "            if line_gene[1] in two_hits['gene'].values:\n",
    "                comp_het_re = re.search(regex, line)\n",
    "                parent1 = [genotype[1], genotype[2]]\n",
    "                parent2 = [genotype[3], genotype[4]]\n",
    "                if genotype[5] in parent1 or genotype[6] in parent1 and genotype[1] != genotype[2] and genotype[5] != genotype[6]:\n",
    "                    comp_het.write(line)\n",
    "                elif genotype[5] in parent2 or genotype[6] in parent2 and genotype[3] != genotype[4] and genotype[5] != genotype[6]:\n",
    "                    comp_het.write(line)\n",
    "            else:\n",
    "                genotype = re.search(regex, line)\n",
    "                if genotype[5] == genotype[6] and float(genotype[5]) > 0:\n",
    "                    parent1 = [genotype[1], genotype[2]]\n",
    "                    parent2 = [genotype[3], genotype[4]]\n",
    "                    if genotype[5] in parent1 and genotype[5] in parent2 and genotype[1] != genotype[2] and genotype[3] != genotype[4]:\n",
    "                        aut_rec.write(line)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "gq_regex = \"EFF=\\S+\\s+GT:AD:DP:GQ\\S*\\s+\\S+\\s+\\S+\\s+\\S*?[:]\\S*?[:]\\S*?[:](.+?)[:]\"\n",
    "## filter with GQ\n",
    "\n",
    "\n",
    "with open('/Users/johnlee/jupyter/new_case/case_study_trio/comp_het.no_null.high_mod.qual100.C4RCD_0637_1_PB_Whole_F0020G2S2M0635P0636_TSE61-C4RCD_0635_1_PB_Whole_F0020G2S1MP_TSE61-C4RCD_0636_1_PB_Whole_F0020G1S1MP_TSE61.phaseBT.vcf', 'rt') as aut_rec, open('/Users/johnlee/jupyter/new_case/case_study_trio/gq99.comp_het.no_null.high_mod.qual100.C4RCD_0637_1_PB_Whole_F0020G2S2M0635P0636_TSE61-C4RCD_0635_1_PB_Whole_F0020G2S1MP_TSE61-C4RCD_0636_1_PB_Whole_F0020G1S1MP_TSE61.phaseBT.vcf', 'w') as output:\n",
    "    for line in aut_rec:\n",
    "        if line.startswith('#'):\n",
    "            output.write(line)\n",
    "        else:\n",
    "            gq = re.search(gq_regex, line)\n",
    "            if float(gq[1]) >= 99:\n",
    "                output.write(line)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "with open('/Users/johnlee/jupyter/new_case/case_study_trio/gq99.comp_het.no_null.high_mod.qual100.C4RCD_0637_1_PB_Whole_F0020G2S2M0635P0636_TSE61-C4RCD_0635_1_PB_Whole_F0020G2S1MP_TSE61-C4RCD_0636_1_PB_Whole_F0020G1S1MP_TSE61.phaseBT.vcf', 'rt') as vcf:\n",
    "    gene_list_regex = \"EFF=\\S*?[|]\\S*?[|]\\S*?[|]\\S*?[|]\\S*?[|](.*?)[|]\"\n",
    "    gene_list = []\n",
    "## MAEK A GENE LIST OF ALL THE GENES > 2 VARIANTS IN THE VCF. SEARCH THESE IN THE RNA SEQ AND LOOK FOR ANY POSSIBLE GENE CANDIDATES\n",
    "    for line in vcf:\n",
    "        if not line.startswith('#'):\n",
    "            gene_name = re.search(gene_list_regex, line)\n",
    "            gene_list.append(gene_name[1])\n",
    "    \n",
    "    genes_df = pd.DataFrame(gene_list)\n",
    "    gene_counts = genes_df[0].value_counts().reset_index()\n",
    "    gene_counts.columns = ['gene', 'count']\n",
    "    two_hits = gene_counts.loc[gene_counts['count'] >1]\n",
    "\n",
    "with open('comp_het.gene_list.txt', 'w') as output:\n",
    "    for x in two_hits['gene']:\n",
    "        output.write(x+ \"\\n\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "### okay, checking the deseq2 results for any effect on the rna. i have a list of compound heterozzygous, as well as autosomal recessive gene candidates. lets get the lines out of the vcf\n",
    "\n",
    "## compound heterozygous; filter for < -1 log2fc\n",
    "with open(\"/Users/johnlee/jupyter/new_case/case_study_trio/sourcefiles/C4RCD_0637-VS-UNAFFECTED.DESeq2_results.txt.deseq.vcf\", 'rt') as vcf, open(\"/Users/johnlee/jupyter/new_case/case_study_trio/comp_het.gene_list.txt\", \"rt\") as gene_list, open('comp_het.genes.vcf', 'w') as output:\n",
    "    genes = []\n",
    "    for line in gene_list:\n",
    "        genes.append(line.strip())\n",
    "    for line in vcf:\n",
    "        if line.startswith('#'):\n",
    "            output.write(line)\n",
    "        else:\n",
    "            regex = \"GENE=(.*?)[;]\\S*[;]LOG2FC=(.*?)[;](.*?);\"\n",
    "            gene_name = re.search(regex, line)\n",
    "            if gene_name[2] != \"NA\":                \n",
    "                if gene_name[1] in genes and float(gene_name[2]) <= -1:\n",
    "                    output.write(line)\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "## compound heterozygous\n",
    "with open(\"/Users/johnlee/jupyter/new_case/case_study_trio/sourcefiles/C4RCD_0637-VS-UNAFFECTED.DESeq2_results.txt.deseq.vcf\", 'rt') as vcf, open(\"/Users/johnlee/jupyter/new_case/case_study_trio/aut_rec.gene_list.txt\", \"rt\") as gene_list, open('aut_rec.genes.vcf', 'w') as output:\n",
    "    genes = []\n",
    "    for line in gene_list:\n",
    "        genes.append(line.strip())\n",
    "    for line in vcf:\n",
    "        if line.startswith('#'):\n",
    "            output.write(line)\n",
    "        else:\n",
    "            regex = \"GENE=(.*?)[;]\\S*[;]LOG2FC=(.*?)[;](.*?);\"\n",
    "            gene_name = re.search(regex, line)\n",
    "            if gene_name[2] != \"NA\":\n",
    "\n",
    "                if gene_name[1] in genes and float(gene_name[2]) <= -1:\n",
    "                    output.write(line)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.5"
  },
  "orig_nbformat": 4
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
